usr/sbin/rear

```diff
@@ -50,23 +68,34 @@ function apply_bash_flags_and_options_commands () {
 #   apply_bash_flags_and_options_commands "$INITIAL_BASH_FLAGS_AND_OPTIONS_COMMANDS"
 readonly INITIAL_BASH_FLAGS_AND_OPTIONS_COMMANDS="$( get_bash_flags_and_options_commands )"
 
-# Enable as needed for development of fail-safe code:
-# set -ue -o pipefail
-# set -xue -o pipefail
-# set -xvue -o pipefail
-# Cf. the Relax-and-Recover Coding Style
-# https://github.com/rear/rear/wiki/Coding-Style
-# that reads (excerpt - dated 18. Nov. 2015):
-# "TODO Use set -eu to die from unset variables and unhandled errors"
-
-# Versioning
-readonly PRODUCT="Relax-and-Recover"
-readonly PROGRAM=${0##*/}
-readonly VERSION=2.9
-readonly RELEASE_DATE="2025-01-31"
-
-# Where users should report bugs:
-readonly BUG_REPORT_SITE="https://github.com/rear/rear/issues"
+# To be on the safe side forbid careless/insecure sourcing
+# via plain 'source' and '.' from the very beginning
+# see in general https://relax-and-recover.org/documentation/security-architecture
+# and see in particular https://github.com/rear/rear/issues/3259
+# and https://github.com/rear/rear/issues/3259#issuecomment-2385745545
+# and https://github.com/rear/rear/issues/3319#issuecomment-2360546784
+# and https://github.com/rear/rear/issues/3319#issuecomment-2363556217
+function . () {
+    echo -e "ERROR: BUG in $PRODUCT\nForbidden usage of '.' in\n  . $*\nPlease report it at $BUG_REPORT_SITE" >&2
+    exit 1
+}
+# To be on the safe side sourcing via '.' is kept always forbidden in ReaR because
+# it is basically impossible to reliably find all code places where '.' is used for sourcing
+# see https://github.com/rear/rear/issues/3260#issuecomment-2248252572
+# In contrast to sourcing via '.' which is kept forbidden
+# for sourcing via 'source' later (see below) trusted sourcing gets set up
+# by replacing this 'source' forbid function by a 'source' wrapper function
+# in lib/_framework-setup-and-functions.sh which implements trusted sourcing
+# which is then the only allowed way to normally source a file in ReaR.
+# For exceptional cases 'builtin source' can still be used to source files
+# that are known to be trusted (e.g. with hardcoded trusted path).
+# The point is that we can reliably find all code places where 'builtin' is used
+# so we can reliably check such exceptional cases at any time to keep ReaR secure
+# (in particular we can also find where the undesirable form 'builtin .' is used):
+function source () {
```

Comment:
As we replace this `source` implementation later on with the wrapper, what does this definition here protect? Can it be that it only protects the code between this definition and the redefinition later on? So specifically this definition protects the code in the `rear` main script between line 95 and line 554 where we load the `_framework-setup-and-functions.sh` which in turns redefines `source` to be the wrapper.

I'm wondering if this is maybe overkill as I expect us to trust ourselves to not wrongly use `source` here.

Reply:
Yes, it is only there to protect **all code**
until the source wrapper function in
_framework-setup-and-functions.sh is defined.

So it protects not only code in sbin/rear but also
code in default.conf which we 'builtin source' in sbin/rear
and also all code in _framework-setup-and-functions.sh
until the source wrapper function is defined therein
which are more than 1000 lines of code therein
until the source wrapper function is defined.

Yes, I don't trust ourselves or any contributor
to never use source by accident in the next 10 years
but then this BUG error shows up immediately
so the one who uses source by accident
gets notified already when he tests his things.